### PR TITLE
spring boot role

### PR DESCRIPTION
**Pull request by Yam0r**

### CODE REVIEW

File: src/main/java/mate/academy/springintro/config/SecurityConfig.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+import org.springframework.security.config.annotation.method.configuration.EnableMethodSecurity;
+import org.springframework.security.config.annotation.web.builders.HttpSecurity;
+import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
+import org.springframework.security.config.annotation.web.configurers.AbstractHttpConfigurer;
+import org.springframework.security.core.userdetails.UserDetailsService;
+import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
+import org.springframework.security.crypto.password.PasswordEncoder;
+import org.springframework.security.web.SecurityFilterChain;
+
+@EnableWebSecurity
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springintro/config/SecurityConfig.java]:
```suggestion
```

File: src/main/java/mate/academy/springintro/config/SecurityConfig.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+@EnableMethodSecurity
+@Configuration
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springintro/config/SecurityConfig.java:18]:
```suggestion
@Configuration
@RequiredArgsConstructor
```

File: src/main/java/mate/academy/springintro/config/SecurityConfig.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+public class SecurityConfig {
+
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springintro/config/SecurityConfig.java]:
```suggestion
```

File: src/main/java/mate/academy/springintro/config/SecurityConfig.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    public SecurityFilterChain securityFilterChain(HttpSecurity http) throws Exception {
+        return http
+                .cors(AbstractHttpConfigurer::disable)
+                .csrf(AbstractHttpConfigurer::disable)
+                .authorizeHttpRequests(
+                        auth -> auth
+                                .requestMatchers("/auth/**", "/error",
+                                        "/swagger-ui/**", "/v3/api-docs/**")
+                                .permitAll()
+                                .requestMatchers("/api/books").hasRole("USER")
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springintro/config/SecurityConfig.java]: don't complicate codem let's use annotations unders the controllers endpoints

File: src/main/java/mate/academy/springintro/model/User.java [hunk 5]
```diff
@@ -2,34 +2,87 @@
...
+
     @Column(nullable = false)
     private String firstName;
+
     @Column(nullable = false)
     private String lastName;
+
     private String shippingAddress;
+
+    @ManyToMany(fetch = FetchType.EAGER)
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springintro/model/User.java]:
```suggestion
    @ManyToMany
```

File: src/main/java/mate/academy/springintro/model/User.java [hunk 6]
```diff
@@ -2,34 +2,87 @@
...
+    @JoinTable(
+            name = "users_roles",
+            joinColumns = @JoinColumn(name = "user_id"),
+            inverseJoinColumns = @JoinColumn(name = "role_id")
+    )
+    private Set<Role> roles;
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springintro/model/User.java]:
```suggestion
    private Set<Role> roles = new HashSet<>();
```

File: src/main/java/mate/academy/springintro/rolerepository/Role.java [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    @Column(nullable = false, unique = true)
+    @Enumerated(EnumType.STRING)
+    private RoleName role;
+
+    public Role(RoleName role) {
+        this.role = role;
+    }
+
+    public Role() {
+    }
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springintro/rolerepository/Role.java]:
```suggestion
```

File: src/main/java/mate/academy/springintro/rolerepository/Role.java [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+import jakarta.persistence.Entity;
+import jakarta.persistence.EnumType;
+import jakarta.persistence.Enumerated;
+import jakarta.persistence.GeneratedValue;
+import jakarta.persistence.GenerationType;
+import jakarta.persistence.Id;
+import jakarta.persistence.Table;
+import org.springframework.security.core.GrantedAuthority;
+
+@Entity
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springintro/rolerepository/Role.java:15]:
```suggestion
@Entity
@Getter
@Setter
```

File: src/main/java/mate/academy/springintro/security/CustomUserDetailsService.java [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+@Service
+@RequiredArgsConstructor
+public class CustomUserDetailsService implements UserDetailsService {
+    private static final String NOT_FOUND_EMAIL_MESSAGE = "Can't find user with email:";
+    private final UserRepository userRepository;
+
+    @Override
+    public UserDetails loadUserByUsername(String email) throws UsernameNotFoundException {
+        return userRepository.findByEmail(email)
+                .orElseThrow(() -> new UsernameNotFoundException(NOT_FOUND_EMAIL_MESSAGE));
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springintro/security/CustomUserDetailsService.java]:
```suggestion
                .orElseThrow(() -> new EntityNotFoundException(NOT_FOUND_EMAIL_MESSAGE));
```

File: src/main/java/mate/academy/springintro/service/impl/UserServiceImpl.java [hunk 10]
```diff
@@ -1,28 +1,43 @@
...
 
 @Service
 @RequiredArgsConstructor
 public class UserServiceImpl implements UserService {
     private final UserRepository userRepository;
     private final UserMapper userMapper;
+    private final PasswordEncoder passwordEncoder;
+    private final RoleRepository roleRepository;
 
     @Override
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springintro/service/impl/UserServiceImpl.java:30]:
```suggestion
    @Override
    @Transactional
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/mate/academy/springintro/config/SecurityConfig.java [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+import org.springframework.security.config.annotation.web.configurers.AbstractHttpConfigurer;
+import org.springframework.security.core.userdetails.UserDetailsService;
+import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
+import org.springframework.security.crypto.password.PasswordEncoder;
+import org.springframework.security.web.SecurityFilterChain;
+
+@EnableMethodSecurity
+@Configuration
+public class SecurityConfig {
+
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springintro/config/SecurityConfig.java]: remove empty line after class declaration

File: src/main/java/mate/academy/springintro/config/SecurityConfig.java [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+import org.springframework.security.config.annotation.method.configuration.EnableMethodSecurity;
+import org.springframework.security.config.annotation.web.builders.HttpSecurity;
+import org.springframework.security.config.annotation.web.configurers.AbstractHttpConfigurer;
+import org.springframework.security.core.userdetails.UserDetailsService;
+import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
+import org.springframework.security.crypto.password.PasswordEncoder;
+import org.springframework.security.web.SecurityFilterChain;
+
+@EnableMethodSecurity
+@Configuration
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springintro/config/SecurityConfig.java:18]: add `@RequiredArgsConstructor`

File: src/main/java/mate/academy/springintro/config/SecurityConfig.java [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+public class SecurityConfig {
+
+    private final UserDetailsService userDetailsService;
+
+    public SecurityConfig(UserDetailsService userDetailsService) {
+        this.userDetailsService = userDetailsService;
+    }
+
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springintro/config/SecurityConfig.java]:
```suggestion
```

File: src/main/java/mate/academy/springintro/model/User.java [hunk 14]
```diff
@@ -5,31 +5,84 @@
...
+import org.hibernate.annotations.Where;
+import org.springframework.security.core.GrantedAuthority;
+import org.springframework.security.core.userdetails.UserDetails;
 
 @Getter
 @Setter
 @Entity
 @SQLDelete(sql = "UPDATE users SET is_deleted = TRUE WHERE id = ?")
-@SQLRestriction(value = "is_deleted = FALSE")
+@Where(clause = "is_deleted = FALSE")
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springintro/model/User.java]: Check if `@Where` is deprecated or not, if yes revert changes

File: src/main/java/mate/academy/springintro/repository/UserRepository.java [hunk 15]
```diff
@@ -1,8 +1,11 @@
 package mate.academy.springintro.repository;
 
+import java.util.Optional;
 import mate.academy.springintro.model.User;
 import org.springframework.data.jpa.repository.JpaRepository;
 
 public interface UserRepository extends JpaRepository<User, Long> {
     boolean existsByEmail(String email);
+
+    Optional<User> findByEmail(String email);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springintro/repository/UserRepository.java:12]:
```suggestion
    @EntityGraph(attributePaths = "roles")
    Optional<User> findByEmail(String email);
```

File: src/main/java/mate/academy/springintro/rolerepository/Role.java [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+    @Column(nullable = false, unique = true)
+    @Enumerated(EnumType.STRING)
+    private RoleName role;
+
+    public Role(RoleName role) {
+        this.role = role;
+    }
+
+    public Role() {
+    }
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springintro/rolerepository/Role.java]: remove constructors
```suggestion
```

File: src/main/java/mate/academy/springintro/rolerepository/Role.java [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+    public String getAuthority() {
+        return role.name();
+    }
+
+    public enum RoleName {
+        USER,
+        ADMIN
+    }
+}
+
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springintro/rolerepository/Role.java]: remove redundant empty line

File: src/main/java/mate/academy/springintro/service/impl/UserServiceImpl.java [hunk 18]
```diff
@@ -1,28 +1,48 @@
...
-            throw new RegistrationException("Can't register user");
+            throw new RegistrationException(NOT_REGISTRATION_EMAIL_MESSAGE);
         }
+
         User user = userMapper.toUser(requestDto);
+        user.setPassword(passwordEncoder.encode(requestDto.getPassword()));
+
+        Role userRole = roleRepository.findByRole(Role.RoleName.USER)
+                .orElseThrow(() ->
+                        new IllegalStateException(NOT_FOUND_ROLE));
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springintro/service/impl/UserServiceImpl.java]:
```suggestion
                        new EntityNotFoundException(NOT_FOUND_ROLE));
```

File: src/main/java/mate/academy/springintro/service/impl/UserServiceImpl.java [hunk 19]
```diff
@@ -1,28 +1,48 @@
...
     public UserResponseDto register(UserRegistrationRequestDto requestDto)
             throws RegistrationException {
         if (userRepository.existsByEmail(requestDto.getEmail())) {
-            throw new RegistrationException("Can't register user");
+            throw new RegistrationException(NOT_REGISTRATION_EMAIL_MESSAGE);
         }
+
         User user = userMapper.toUser(requestDto);
+        user.setPassword(passwordEncoder.encode(requestDto.getPassword()));
+
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springintro/service/impl/UserServiceImpl.java:40]: remove redundant empty lines in method

File: src/main/resources/db/changelog/db.changelog-master.yaml [hunk 20]
```diff
@@ -5,3 +5,13 @@ databaseChangeLog:
...
+  - include:
+      file: db/changelog/changes/set-bcrypt-password.yaml
+  - include:
+      file: db/changelog/changes/create-roles-table.yaml
+  - include:
+      file: db/changelog/changes/create-users-roles-table.yaml
+  - include:
+      file: db/changelog/changes/insert-roles.yaml
+  - include:
+      file: db/changelog/changes/insert-roles-of-users.yaml
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/db.changelog-master.yaml]: add assign-roles-to-users changeset

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/mate/academy/springintro/rolerepository/Role.java [hunk 21]
```diff
@@ -0,0 +1,37 @@
+package mate.academy.springintro.rolerepository;
```
Comment (ahoienko) [inline src/main/java/mate/academy/springintro/rolerepository/Role.java:1]: What is the reason to have a separate package for role? Is it possible to add Role to model package and RoleRepository to repository package?

File: src/main/java/mate/academy/springintro/security/CustomUserDetailsService.java [hunk 22]
```diff
@@ -0,0 +1,22 @@
...
+@Service
+@RequiredArgsConstructor
+public class CustomUserDetailsService implements UserDetailsService {
+    private static final String NOT_FOUND_EMAIL_MESSAGE = "Can't find user with email:";
+    private final UserRepository userRepository;
+
+    @Override
+    public UserDetails loadUserByUsername(String email) throws UsernameNotFoundException {
+        return userRepository.findByEmail(email)
+                .orElseThrow(() -> new EntityNotFoundException(NOT_FOUND_EMAIL_MESSAGE));
```
Comment (ahoienko) [inline src/main/java/mate/academy/springintro/security/CustomUserDetailsService.java]:
```suggestion
                .orElseThrow(() -> new EntityNotFoundException(NOT_FOUND_EMAIL_MESSAGE + email));
```

File: src/main/java/mate/academy/springintro/service/impl/UserServiceImpl.java [hunk 23]
```diff
@@ -1,28 +1,48 @@
...
         if (userRepository.existsByEmail(requestDto.getEmail())) {
-            throw new RegistrationException("Can't register user");
+            throw new RegistrationException(NOT_REGISTRATION_EMAIL_MESSAGE);
         }
+
         User user = userMapper.toUser(requestDto);
+        user.setPassword(passwordEncoder.encode(requestDto.getPassword()));
+        Role userRole = roleRepository.findByRole(Role.RoleName.USER)
+                .orElseThrow(() ->
+                        new EntityNotFoundException(NOT_FOUND_ROLE));
```
Comment (ahoienko) [inline src/main/java/mate/academy/springintro/service/impl/UserServiceImpl.java]:
```suggestion
                        new EntityNotFoundException(NOT_FOUND_ROLE + roleName));
```

**Review by ahoienko — CHANGES_REQUESTED** [review]
Nice, left minor comments

File: src/main/resources/db/changelog/db.changelog-master.yaml [hunk 24]
```diff
@@ -5,3 +5,15 @@ databaseChangeLog:
...
+  - include:
+      file: db/changelog/changes/create-roles-table.yaml
+  - include:
+      file: db/changelog/changes/create-users-roles-table.yaml
+  - include:
+      file: db/changelog/changes/insert-roles.yaml
+  - include:
+      file: db/changelog/changes/insert-roles-of-users.yaml.
+  - include:
+      file: db/changelog/changes/assign-roles-to-users.yaml
```
Comment (okuzan) [inline src/main/resources/db/changelog/db.changelog-master.yaml:19]: image
don't forget to include an empty line at the EOF

**Review by okuzan — APPROVED** [review]

**Pull request merged**